uefi-test-runner/src/boot/memory.rs

```diff
@@ -37,7 +38,7 @@ fn allocate_pages(bt: &BootServices) {
     buf[4095] = 0x23;
 
     // Clean up to avoid memory leaks.
-    bt.free_pages(pgs, 1).unwrap();
+    bt.free_pages(pgs, 1).unwrap().unwrap();
```

Comment:
Hmm, could we somehow keep two `Results`, one for which warnings are meaningful, and one for which Warnings are treated as errors?

I feel like this function shouldn't fail (which is why `unwrap`) was used in the first place, but even if a warning is returned because the firmware thought so it should be treated as an error (since no meaningful action could be taken anyway).

Reply:
Short version: what I would like to see is something which looks like a Result, but has extra methods for convenient warning handling like Completion currently has: one method which unwraps both the outer Result and the inner Completion, panicking if there are either errors or warnings, one method which panics if there are errors but which only logs the warnings...

<details>
<summary>Long version with extra API design thoughts</summary>
I think the core `uefi-rs` crates should, by default, follow UEFI's error handling decisions. If UEFI says that an error is fatal, we abort immediately. If UEFI says that an error is not fatal (a "warning"), we continue trying to go as far as we can and report the error at the end.

There are cases where we deviate from this rule, because UEFI exposes as an error something which we know to actually be a non-fatal failure (e.g. NOT_READY in the text input protocol, TIMEOUT in serial communication). In this case, we downgrade the error into a warning or an expected result as appropriate.

However, I would be wary of doing the reverse (strengthening a regular result or a warning into a fatal error) in the core `uefi-rs` crate. It seems to me that this is a choice which user code like `uefi-test-runner` should be making, not the core library itself.

Neveretheless, a test harness which treats warnings as errors like `uefi-test-runner` does should be easier to write than it currently is in this PR.
</details>